### PR TITLE
coord: make mz_catalog_names a view

### PR DESCRIPTION
**Pull request by benesch**
Now that we track tables, sources, etc. in their own tables,
mz_catalog_names can be a simple view over mz_objects. This saves a good
bit of bookkeeping.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4366)

### CODE REVIEW

File: src/transform/src/dataflow.rs [hunk 1]
```diff
@@ -23,7 +23,7 @@ use std::collections::{HashMap, HashSet};
 /// This method is currently limited in scope to propagating filtering and
 /// projection information, though it could certainly generalize beyond.
 pub fn optimize_dataflow(dataflow: &mut DataflowDesc) {
-    optimize_dataflow_filters(dataflow);
```
Comment (benesch) [inline src/transform/src/dataflow.rs:26]: Mentioned on Slack; without this one of the queries in show.slt is misoptimized. Need to resolve this before merging.

Comment (benesch) [reply src/transform/src/dataflow.rs]: Fixed by #4362.

File: src/coord/src/catalog/builtin.rs [hunk 2]
```diff
@@ -361,11 +353,64 @@ lazy_static! {
...
+// For historical reasons, this view does not properly escape identifiers. For
+// example, a table named 'cAp.S' in the default database and schema will be
+// rendered as `materialize.public.cAp.S`. This is *not* a valid SQL identifier,
+// as it has an extra dot, and the capitals will be folded to lowercase. This
+// view is thus only fit for use in debugging queries where the names are for
+// human consumption. Applications should instead pull the names of individual
+// components out of mz_objects, mz_schemas, and mz_databases to avoid these
+// escaping issues.
+//
+// TODO(benesch): deprecate this view.
```
Comment (JLDLaughlin) [inline src/coord/src/catalog/builtin.rs:396]: What do you want to replace this with in the future?

**Review by JLDLaughlin — APPROVED** [review]
🎉 awesome!

File: src/coord/src/catalog/builtin.rs (hunk 2 above)
Comment (benesch) [reply src/coord/src/catalog/builtin.rs:396]: Ideally nothing! The names of things are in `mz_objects` already, so most queries can just be written to use the name from there. If you do need the fully qualified name, joining against `mz_schemas` and `mz_databases` means you'll get the database, schema, and item name in three separate columns, which is much more reliable.

**Pull request merged**